### PR TITLE
Fix AWS SQS bug

### LINKED ISSUE #86: Fails to apply a plan with aws_sqs_queue_policy defined
##### SUMMARY

I have a terraform plan that defines an `aws_sqs_queue_policy` resource.  Using the cloud.terraform.terraform module to apply the plan fails with a Python exception.  If I comment out or remove the `aws_sqs_queue_policy` resource, applying the plan succeeds as expected.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

`cloud.terraform.terraform` module

##### ANSIBLE VERSION

```paste below
$ ansible --version
ansible [core 2.14.4]
  config file = /home/localadmin/dev/reproduce-ansible-terraform-error/ansible.cfg
  configured module search path = ['/home/localadmin/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  ansible collection location = /home/localadmin/dev/reproduce-ansible-terraform-error/collections
  executable location = /usr/bin/ansible
  python version = 3.9.14 (main, Jan  9 2023, 00:00:00) [GCC 11.3.1 20220421 (Red Hat 11.3.1-2)] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION

```paste below
$ ansible-galaxy collection list cloud.terraform

# /home/localadmin/dev/reproduce-ansible-terraform-error/collections/ansible_collections
Collection      Version
--------------- -------
cloud.terraform 1.1.1
```

##### CONFIGURATION

```paste below
$ ansible-config dump --only-changed
COLLECTIONS_PATHS(/home/user/dev/reproduce-ansible-terraform-error/ansible.cfg) = ['/home/user/dev/reproduce-ansible-terraform-error/collections']
CONFIG_FILE() = /home/user/dev/reproduce-ansible-terraform-error/ansible.cfg
```

##### OS / ENVIRONMENT

OS is RHEL 9.1

```
$ terraform --version
Terraform v1.4.2
on linux_amd64

Your version of Terraform is out of date! The latest version
is 1.5.4. You can update by downloading from https://www.terraform.io/downloads.html
```

##### STEPS TO REPRODUCE

I'm using the following main.tf:

```
provider "aws" {
  region = "us-east-2"
}

resource "aws_sqs_queue" "this" {
  name = "reproducer-queue"
}

data "aws_iam_policy_document" "this" {
  statement {
    effect = "Allow"

    principals {
      type = "AWS"
      identifiers = [
        aws_iam_role.this.arn
      ]
    }

    actions = [
      "sqs:SendMessage"
    ]

    resources = [
      aws_sqs_queue.this.arn
    ]
  }
}

resource "aws_sqs_queue_policy" "this" {
  queue_url = aws_sqs_queue.this.id
  policy    = data.aws_iam_policy_document.this.json
}

resource "aws_iam_role" "this" {
  name = "eda_forwarder_role"

  assume_role_policy = jsonencode({
    Version = "2012-10-17"
    Statement = [{
      Action = "sts:AssumeRole"
      Effect = "Allow"
      Sid    = ""
      Principal = {
        Service = "lambda.amazonaws.com"
      }
    }]
  })
}

resource "aws_iam_role_policy_attachment" "this" {
  role       = aws_iam_role.this.name
  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
}
```

This playbook is used to apply the terraform plan:

```yaml
---
- name: Reproduce cloud.terraform apply error
  hosts: localhost

  tasks:
    - name: Apply terraform AAP infrastructure plan
      cloud.terraform.terraform:
        project_path: '{{ playbook_dir }}/build'
        state: present
        force_init: true

...
```

##### EXPECTED RESULTS

Playbook should succeed and all defined terraform resources should be created.

##### ACTUAL RESULTS

```paste below
$ ansible-playbook apply.yml -vvv
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'
ansible-playbook [core 2.14.4]
  config file = /home/user/dev/reproduce-ansible-terraform-error/ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  ansible collection location = /home/user/dev/reproduce-ansible-terraform-error/collections
  executable location = /usr/bin/ansible-playbook
  python version = 3.9.14 (main, Jan  9 2023, 00:00:00) [GCC 11.3.1 20220421 (Red Hat 11.3.1-2)] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
Using /home/user/dev/reproduce-ansible-terraform-error/ansible.cfg as config file
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Parsed /etc/ansible/hosts inventory source with ini plugin
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: apply.yml ************************************************************
1 plays in apply.yml

PLAY [Reproduce cloud.terraform apply error] ***********************************

TASK [Gathering Facts] *********************************************************
task path: /home/user/dev/reproduce-ansible-terraform-error/apply.yml:2
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: user
<127.0.0.1> EXEC /bin/sh -c 'echo ~user && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user/.ansible/tmp `"&& mkdir "` echo /home/user/.ansible/tmp/ansible-tmp-1690921267.7795496-12710-197576117
063153 `" && echo ansible-tmp-1690921267.7795496-12710-197576117063153="` echo /home/user/.ansible/tmp/ansible-tmp-1690921267.7795496-12710-197576117063153 `" ) && sleep 0'
Using module file /usr/lib/python3.9/site-packages/ansible/modules/setup.py
<127.0.0.1> PUT /home/user/.ansible/tmp/ansible-local-12705g5zbt519/tmp0pmn7il9 TO /home/user/.ansible/tmp/ansible-tmp-1690921267.7795496-12710-197576117063153/AnsiballZ_set
up.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/user/.ansible/tmp/ansible-tmp-1690921267.7795496-12710-197576117063153/ /home/user/.ansible/tmp/ansible-tmp-1690921267.7795496-1
2710-197576117063153/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /home/user/.ansible/tmp/ansible-tmp-1690921267.7795496-12710-197576117063153/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/user/.ansible/tmp/ansible-tmp-1690921267.7795496-12710-197576117063153/ > /dev/null 2>&1 && sleep 0'
ok: [localhost]

TASK [Apply terraform AAP infrastructure plan] *********************************                                                                                    [34/1417]
task path: /home/user/dev/reproduce-ansible-terraform-error/apply.yml:6
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: user
<127.0.0.1> EXEC /bin/sh -c 'echo ~user && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user/.ansible/tmp `"&& mkdir "` echo /home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-16757231678
8460 `" && echo ansible-tmp-1690921268.38882-12787-167572316788460="` echo /home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460 `" ) && sleep 0'
Using module file /home/user/dev/reproduce-ansible-terraform-error/collections/ansible_collections/cloud/terraform/plugins/modules/terraform.py
<127.0.0.1> PUT /home/user/.ansible/tmp/ansible-local-12705g5zbt519/tmp8q5xr8cn TO /home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/AnsiballZ_terra
form.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/ /home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787
-167572316788460/AnsiballZ_terraform.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/AnsiballZ_terraform.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/AnsiballZ_terraform.py", line 107, in <module>
    _ansiballz_main()
  File "/home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/AnsiballZ_terraform.py", line 99, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/AnsiballZ_terraform.py", line 47, in invoke_module
    runpy.run_module(mod_name='ansible_collections.cloud.terraform.plugins.modules.terraform', init_globals=dict(_module_fqn='ansible_collections.cloud.terraform.plugins.mod
ules.terraform', _modlib_path=modlib_path),
  File "/usr/lib64/python3.9/runpy.py", line 225, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib64/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/modules/terraform.p
y", line 659, in <module>
  File "/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/modules/terraform.p
y", line 591, in main
  File "/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/terraf
orm_commands.py", line 215, in show
  File "/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models
.py", line 108, in from_json
  File "/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models
.py", line 93, in from_json
  File "/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models
.py", line 78, in from_json
  File "/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models
.py", line 78, in <listcomp>
  File "/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models
.py", line 49, in from_json
KeyError: 'values'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/AnsiballZ_terraform.py\", line
107, in <module>\n    _ansiballz_main()\n  File \"/home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/AnsiballZ_terraform.py\", line 99, in _ansiballz
_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/user/.ansible/tmp/ansible-tmp-1690921268.38882-12787-167572316788460/AnsiballZ_terraform.py
\", line 47, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.cloud.terraform.plugins.modules.terraform', init_globals=dict(_module_fqn='ansible_collecti
ons.cloud.terraform.plugins.modules.terraform', _modlib_path=modlib_path),\n  File \"/usr/lib64/python3.9/runpy.py\", line 225, in run_module\n    return _run_module_code(co
de, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.9/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/
lib64/python3.9/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terr
aform_payload.zip/ansible_collections/cloud/terraform/plugins/modules/terraform.py\", line 659, in <module>\n  File \"/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0
/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/modules/terraform.py\", line 591, in main\n  File \"/tmp/ansible_cloud.terraform.t
erraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/terraform_commands.py\", line 215, in show\n
  File \"/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/model
s.py\", line 108, in from_json\n  File \"/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terr
aform/plugins/module_utils/models.py\", line 93, in from_json\n  File \"/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip
/ansible_collections/cloud/terraform/plugins/module_utils/models.py\", line 78, in from_json\n  File \"/tmp/ansible_cloud.terraform.terraform_payload_dc9y5y_0/ansible_cloud.
terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models.py\", line 78, in <listcomp>\n  File \"/tmp/ansible_cloud.terraform.terraform
_payload_dc9y5y_0/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models.py\", line 49, in from_json\nKeyError: 'value
s'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

### PR DESCRIPTION
**Pull request by justinc1**
##### SUMMARY

JSON data received from terraform might not contain `values` field. This prevents working with AWS SQS.

Sample code from #86 is included into an integration test.
Small complication - the `amazon.aws.iam_role_info` is not yet released, so I did run tests with amazon.aws from git master.
The check using `iam_role_info` can be also commented out, and only success/failure of `cloud.terraform.terraform` is checked.

Fixes #86

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

terraform

##### ADDITIONAL INFORMATION

Unit test for json data with/without `values` is included. The sample data was logged by adding `q(json.dumps(json_data))` to code, and running integration test. So sample data is realistic.

### CODE REVIEW

File: tests/integration/targets/aws_sqs_queue/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      vars:
+        number_of_roles: 0
+
+  # Clean up all integration test resources
+  always:
+    - name: Fetch IAM role info
+      amazon.aws.iam_role_info:
+        name: "{{ test_role_name }}"
+      register: role_info
+      ignore_errors: true
```
Comment (gravesm) [inline tests/integration/targets/aws_sqs_queue/tasks/main.yml:147]: This shouldn't be necessary. There would be only one role with this name so you can just delete the role by name without the info task.

File: tests/integration/targets/aws_sqs_queue/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+    - name: Delete IAM role
+      amazon.aws.iam_role:
+        name: "{{ item.role_name }}"
+        state: absent
+      loop: "{{ role_info.iam_roles }}"
+      ignore_errors: true
+
+    # TODO remove SQS, when amazon.aws has module for SQS
```
Comment (gravesm) [inline tests/integration/targets/aws_sqs_queue/tasks/main.yml:156]: You should be able to use the `community.aws.sqs_queue` module here.

File: tests/integration/targets/aws_sqs_queue/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        - name: Assert that there are {{ number_of_roles }} roles with role_name={{ test_role_name }}'
+          assert:
+            that:
+              - (role_info.iam_roles | length) == number_of_roles
+        # TODO: uncomment when iam_role_info does return tags
+        # - name: Assert that role {{ test_role_name }} is present and the info matches
+        #  assert:
+        #    that:
+        #      - role_info.iam_roles[0].tags.Name == "{{ resource_id }}"
+        #  when: number_of_roles == 1
```
Comment (gravesm) [inline tests/integration/targets/aws_sqs_queue/tasks/main.yml:42]: The amazon.aws 7.0.0 release is planned for Nov 2, but IMO we don't need this level of verification for this collection. The goal should be to test that terraform runs correctly and the module doesn't blow up. Definitely worth having a test for this bug, but you can just use the `cloud.terraform` module to do that.

**Review by gravesm — COMMENTED** [review]
Thanks for the PR. A few things. Also, this will need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#community-changelogs).

File: tests/integration/targets/aws_sqs_queue/tasks/main.yml (hunk 3 above)
Comment (justinc1) [reply tests/integration/targets/aws_sqs_queue/tasks/main.yml:42]: OK, removed dependency on amazon.aws 7.0.0. The cleanup task for role uses `amazon.aws.iam_role`, so I commented it out. It is not really needed, terraform with state=absent does cleanup.

**Comment by justinc1** [comment]
I'm not sure if you @gravesm prefer to have changes in separate commit, or should I squash early. I pushed a separate commit.

**Comment by gravesm** [comment]
> I'm not sure if you @gravesm prefer to have changes in separate commit, or should I squash early. I pushed a separate commit.

Separate commits are preferred. They'll get squashed when the PR is merged. Overall, this looks good. Once amazon.aws is released next week, we can make sure the tests are passing and get this merged.

**Comment by justinc1** [comment]
The integration test does not depend on amazon.aws>=7.0.0 anymore. It does pass when I run it locally.

**Comment by justinc1** [comment]
I tried to fix linter and sanity tests, but seems linter (`tox  -e linters -vv --skip-missing-interpreters=false`) will want single line

```
"tags": {"Name": "ansible-test-59219564-jcpc-sqs", "cloud_terraform_integration": "true"},
```

But sanity test `ansible-test sanity -v --color --coverage --junit --docker --python 3.10` will want
```
"tags": {
    "Name": "ansible-test-59219564-jcpc-sqs",
    "cloud_terraform_integration": "true"
},
```

The integration test `aws_sqs_queue` failed as `couldn't resolve module/action 'community.aws.sqs_queue'. This often indicates a misspelling, missing collection, or incorrect module path`. I thought adding `community.aws` to `tests/integration/requirements.yml` should be used to solve this, but I'm not sure any more.

**Comment by hakbailey** [comment]
> I tried to fix linter and sanity tests, but seems linter (`tox -e linters -vv --skip-missing-interpreters=false`) will want single line

I think adding a trailing comma to the offending dicts will keep the black linter from complaining about multiple lines. For example:
````
"tags": {
    "Name": "ansible-test-59219564-jcpc-sqs",
    "cloud_terraform_integration": "true",
},
```

**Comment by hakbailey** [comment]
> The integration test `aws_sqs_queue` failed as `couldn't resolve module/action 'community.aws.sqs_queue'. This often indicates a misspelling, missing collection, or incorrect module path`. I thought adding `community.aws` to `tests/integration/requirements.yml` should be used to solve this, but I'm not sure any more.

This appears to be an issue with the integration tests CI workflow in this collection, it's hard-coded the dependency installs instead of using `tests/integration/requirements.yml`. I'll update this in #102 and this problem should be resolved once that's merged.

**Comment by hakbailey** [comment]
@justinc1 #102 has been merged, if you rebase onto current main and force push it should resolve the integration test dependency issue.

**Comment by justinc1** [comment]
Rebase done @hakbailey. Thank you.

**Comment by justinc1** [comment]
Ah, I push tiny linter fix, and canceled the run. Sorry @hakbailey ; I wont push anymore today.

**Review by abikouo — APPROVED** [review]

**Review by gravesm — APPROVED** [review]

**Pull request merged**